refactor: introduce lazycompilation-middleware (#9515)
web-infra-dev/rspack: website/docs/en/guide/features/lazy-compilation.mdx

In addition to two coarse-grained configurations `entries` and `imports`, you can also use a `test` configuration to filter out certain modules for lazy compilation. If you want to disable lazy compilation for the `About` entry point, you can refer to the following configuration:

```js title="rspack.config.mjs"
import { defineConfig } from '@rspack/cli';

export default defineConfig({
  entry: {
    Home: './src/Home.js',
    About: './src/About.js',
  },
  experiments: {
    lazyCompilation: {
      entries: true,
      imports: true,
      test(module) {
        const name = module.nameForCondition();
        return name && !/src\/About/.test(name);
      },
    },
  },
});
```

## Under the hood

The principle of lazy compilation is to proxy the unexecuted entries and dynamically imported modules. When the module is executed during runtime it sends a request to the dev server, triggering rebuild by Compiler along with module hot updates.

Only when corresponding entries and modules are executed will Rspack compile their respective entries and Modules along with all their dependencies.

![image](https://assets.rspack.dev/rspack/assets/lazy-proxy-module.png)

> ## Using Custom Development Servers

Reviewer: ```suggestion
## Using custom dev server
```

> In the above configuration, `experiments.lazyCompilation` configuration is actually consumed by `@rspack/cli` and adds an [Express style middleware](https://expressjs.com/en/guide/using-middleware.html) in handling requests related to lazyCompilation from clients into `@rspack/dev-server`. If using your own development server, you should manually add this middleware logic into your development server.

Reviewer: ```suggestion
In the configuration above, the `experiments.lazyCompilation` option is actually processed by `@rspack/cli`. It adds an [Express-style middleware](https://expressjs.com/en/guide/using-middleware.html) to `@rspack/dev-server` specifically designed to handle lazy compilation client requests.

If you are using a custom dev server, you will need to manually integrate this middleware into your dev server.
```

> ```js title="start.js"

Reviewer: ```suggestion
```js title="start.mjs"
```

import { experiments, rspack } from '@rspack/core';
import config from './rspack.config.mjs';
import DevServer from 'webpack-dev-server';

const compiler = rspack(config);

> const middleware = experiments.lazyCompilationMiddleware(

Reviewer: Can we add some descriptions for the middleware options.

Author: The detailed option is right above this.
